### ISSUE TITLE
sinking in liquids doesn't depend on player's gravity

### ISSUE THREAD

**Issue by DS-Minetest** [minetest/minetest#6123]
##### Issue type
 - Feature request
 
##### Summary
- Eg. if you set the player gravity to -1 and are in water, you sink down, not up.
- Perhaps a `liquid_sink_speed` field could also be added to the `override_table` of `set_physics_override`.
- Some realistic usecases: Just think of a suit with that you don't sink to make swimming easier. Or armor could make you more heavy and sink faster. (Or an antigravity node placed underwater.)

**Comment by paramat** [comment]
Negative gravity is possible but not really intended.
I thnk it is not worth taking into account the gravity setting, sink speed is slo slow anyway it is suitable for almost any gravity setting, even a very low one used on a small moon or asteroid.
A fast sink speed is not wanted (we reverted a faster sink in MTG) so should not happen with high gravity settings.

**Comment by Wuzzy2** [comment]
Well, Minetest is not made for *negative* gravities. It needs a lot of work before that works in a consistent (!) manner and not just somehow.

It does seem strange, however, that the water sinking speed is completely decoupled from the gravity. On the other hand, those things can be changed in the settings already, but tweaking could be a bit annoying.

**Comment by paramat** [comment]
So i think it's best kept as it is, a separate and independent setting :-1:

**Comment by DS-Minetest** [comment]
A `liquid_sink_speed` field could also be added to the `override_table` of `set_physics_override`.

**Comment by paramat** [comment]
It could, but pointless because setting it per-player and in-game is not needed.

**Comment by DS-Minetest** [comment]
>setting it per-player and in-game is not needed.

Just think of a suit with that you don't sink to make swimming easier. Or armor could make you more heavy and sink faster. (Or an antigravity node placed underwater.)

**Comment by paramat** [comment]
Yes it's possible to think up a use if you try, but those uses are very specialised and not likely to be used much. We only add physics overrides for very useful things. Being able to imagine a use if you try hard enough is not a justification.

**Comment by MisterXtreme** [comment]
So, we should be able to change normal gravity, but not liquid gravity, due to you thinking the mentioned ideas are "special case"?

**Comment by paramat** [comment]
No, because of all the reasons i explain above. ALso, no-one has needed this yet, people are only imagining uses.

**Comment by paramat** [comment]
1 year bump for more core dev consideration.

**Comment by paramat** [comment]
If gravity was -1, which is unplayable anyway, a liquid should float upwards too, so not really a relevant observation.
> Some realistic usecases: Just think of a suit with that you don't sink to make swimming easier. Or armor could make you more heavy and sink faster.

This would be a bouyancy feature. If done using per-player gravity then gravity out of the water would be affected too which you don't want. So not really a good usecase.

Sinking in a liquid is realistically determined by player weight and density, and liquid density, not by gravity. Whether you go up or down and how fast is determined by the player's density relative to the liquid density, it's independent of gravity. It's also determined by currents which tend to suck you under. I also see no need or usefulness in sinking slower or faster according to per-player gravity as a non-realistic feature.
> Perhaps a liquid_sink_speed field could also be added to the override_table of set_physics_override.

This makes more sense, but something has to be very important to add to physics overrides and this doesn't seem significant enough.

**Comment by numberZero** [comment]
> it's independent of gravity.

WTF? It depends *only* on gravity and density (assuming viscosity is zero, but that’s another story):

 *F_gravity = −m ⋅ g = −ρ_body ⋅ V_body ⋅ g* (assuming the density is constant; integrate otherwise)

 *F_buoyancy = ρ_liquid ⋅ V_body ⋅ g*

 *F_total = F_gravity + F_buoyancy = (ρ_liquid − ρ_body) ⋅ V_body ⋅ g*

 *a = F_total / m = F_total / (ρ_body ⋅ V_body) = (ρ_liquid / ρ_body − 1) ⋅ g*

The use case is realms with different gravity.

**Comment by numberZero** [comment]
Note that I don’t advocate negative gravity.

**Comment by paramat** [comment]
Yes, it's not independent of gravity, my sloppy writing.
I was trying to say, whether you rise or fall is determined by player density and buoyancy, not gravity. So behaviour is already unrealistic.
Ayway i'm now neutral.
SmallJoker i'm assuming you support the request.

**Comment by SmallJoker** [comment]
Yes, water physics should be implemented properly. However, I'm afraid that it might again break the movement code, thus needing adjustments to make it work very similar to the current code.

**Comment by Hume2** [comment]
While you are at it, you can implement liquid densities as suggested here: https://forum.minetest.net/viewtopic.php?f=5&t=22153

**Comment by paramat** [comment]
> water physics should be implemented properly

The request is not 'proper physics' because whether you rise or fall, and how fast, is also determined by player density, liquid density and proportion of player immersed. Not by gravity alone.
We can't and shouldn't simulate realistic physics, however, the key point is:
Vertical speed in liquid cannot be determined from gravity.
> Eg. if you set the player gravity to -1 and are in water, you sink down, not up.

If gravity was 0 or negative, there wouldn't be a liquid surface anyway, all the liquid should be floating around of falling upwards.
> Some realistic usecases: Just think of a suit with that you don't sink to make swimming easier. Or armor could make you more heavy and sink faster. (Or an antigravity node placed underwater.)

These usecases have nothing to do with gravity, these are determined by player density and liquid density, this is irrelevant here.
:-1: for request as stated.

However, maybe you could change the issue to be about implementing buoyancy?

**Comment by DS-Minetest** [comment]
Nah, I think, that would be too complex. Games don't have to be realistic. Just think of the fact that you can set the gravity of a single player independent from where he is. Of course, sinking speed doesn't depend only on gravity, but there is actually no real gravitation but instead some magical force that pulls the water down and another magical force that pulls persons (eg. less strong down or upward). How else would you explain that you still fall down with the same acceleration in singlenode mapgen or that there are no tides and co.?

**Comment by paramat** [comment]
Yes i agree coping with buoyancy and relative densities is going too far. And my point is, if that is going too far then being concerned about gravity is certainly going too far, because buoyancy is the primary consideration.

**Comment by DS-Minetest** [comment]
This is not really about realism. It just feels weird if you fall really slowly and the hit the water and sink rapidly (or even if you fall upwards normally and when you hit the water through inertia, you suddenly sink downward).
Maybe the falling speed in water should just depend on the gravity until the point where you can barely swim against the in-liquid-falling acceleration.

**Comment by paramat** [comment]
> It just feels weird if you fall really slowly and the hit the water and sink rapidly

The water sink speed is not rapid, it's very slow by default. If it's a problem you can make it even slower in settings.
> or even if you fall upwards normally and when you hit the water through inertia, you suddenly sink downward

Inverted gravity is not intended or supported by MT, it should really be disallowed.

This whole issue is invalid as explained. Either we consider relative densities of player and liquid, which you have already disagreed with, or this is closed.

**Comment by DS-Minetest** [comment]
> > It just feels weird if you fall really slowly and the hit the water and sink rapidly
> 
> The water sink speed is not rapid, it's very slow by default. If it's a problem you can make it even slower in settings.

What I meant here is that if the gravity is set very very low, sinking might be faster than falling in air.

> > or even if you fall upwards normally and when you hit the water through inertia, you suddenly sink downward
> 
> Inverted gravity is not intended or supported by MT, it should really be disallowed.

This is why I've set it in parenthesis.
Actually, even though it is "not intended or supported", it's a useful hack that is used by some mods.

> This whole issue is invalid as explained. Either we consider relative densities of player and liquid, which you have already disagreed with, or this is closed.

Do we really need something hyper-realistic here while in other places physics are very unrealistic (eg. liquid flowing dynamics, probably no conservation of momentum when objects collide, per player "gravity", …)?
According to numberZero, sinking acceleration is directly proportional to the gravitational acceleration ("gravity" of the player). Can't we just have that until the sinking speed in the liquid would be larger than, say 0.9 nodes per real second squared?

**Comment by paramat** [comment]
> What I meant here is that if the gravity is set very very low, sinking might be faster than falling in air.

That's highly unlikely, even if so sinking will still be very slow so it won't be a radical increase in speed.
> Do we really need something hyper-realistic here while in other places physics are very unrealistic

No, i'm not requesting it, and this is why being concerned about gravity makes no sense, as that is being even more excessively realistic.
> According to numberZero, sinking acceleration is directly proportional to the gravitational acceleration 

Yes, as a qualified physicist i can confirm, but when considering buoyancy, relative densities of player and liquid has far more effect and is the first thing to consider. if you don't consider densities considering gravity is ridiculous.

**Comment by paramat** [comment]
Changed title to keep request open.

**Comment by paramat** [comment]
Hmm actually it's more complex.
NumberZero's formula is for acceleration not speed:
> (assuming viscosity is zero, but that’s another story)

The acceleration (ignoring viscosity) is proportional to gravity.
But sink speed is the 'terminal velocity' reached determined by the acceleration and hydrodynamic drag of the liquid, which depends on player 'aerodynamics' and liquid viscosity.
Liquid viscosity cannot be ignored because liquids are dense and highly viscous.
Even in very high gravity, the terminal sink speed could still be very slow if the liquid is viscous enough.
So sink speed is not proportional to gravity even if you ignore buoyancy. Liquid viscosity is far more important.

There's no way to approach this request other than a quite complex and realistic way. Is it worth it? I don't think so currently.

**Comment by DS-Minetest** [comment]
Btw. here's related code:
https://github.com/minetest/minetest/blob/ffb17f1c9a203fea6de70159b461f52d104e05b9/src/client/clientenvironment.cpp#L164-L195

Imo it would be enough to change this line:
https://github.com/minetest/minetest/blob/ffb17f1c9a203fea6de70159b461f52d104e05b9/src/client/clientenvironment.cpp#L175
to something like
```cpp
	speed.Y -= rangelim(lplayer->movement_liquid_sink *
		lplayer->physics_override_gravity * dtime_part * 2.0f,
		-0.9f, 0.9f;
```

**Comment by numberZero** [comment]
@DS-Minetest Why `rangelim`?

> The acceleration (ignoring viscosity) is proportional to gravity.

> Liquid viscosity cannot be ignored because liquids are dense and highly viscous.

@paramat As I read the code, at line 175 it computes acceleration for zero viscosity, and then it corrects it to account for actual viscosity. Buoyancy is hidden in `lplayer->movement_liquid_sink` (that effectively treats all liquids as having the same density).

**Comment by paramat** [comment]
Oh yeah, thanks, viscosity has an effect, although in a very hacky non-realistic approximation.

**Comment by DS-Minetest** [comment]
> @DS-Minetest Why `rangelim`?

The rangelim was meant to keep the ability to swim in liquids.
This probably doesn't make sense. I'll edit my comment.

**Comment by DS-Minetest** [comment]
PR  #8631
(It seems like this is a bit more difficult with player controls, todo.)

**Comment by paramat** [comment]
Sorry, i didn't look at this carefully enough. Looking at the code `lplayer->movement_liquid_sink` is actually an acceleration, then viscosity is added to cause a terminal velocity.
So
```
	speed.Y -= lplayer->movement_liquid_sink *
		lplayer->physics_override_gravity * dtime_part * 2.0f;
```
Is simple and reasonable.
I'm neutral, removed my -1.

**Comment by Hume2** [comment]
If all objects were approximated by the same hydrodynamics, I think, it would make it simple enough. I think, there is a reason to implement liquid densities, basically because of objects floating on water or swimming in water. These objects, such as boat or fish, have long codes which only check if they are in water and change their physics accordingly. This all could be omitted if the densities were implemented properly.

Also, have you ever noticed that a human can't sink in lava, not even his dead body?

**Comment by DS-Minetest** [comment]
Densities for objects might indeed be interesting. But Imo they should be discussed in another issue as this one is bloated with other stuff which will slow down productive discussion on densities.
(Actually entities already have an unused `weight` property. (See also https://github.com/minetest/minetest_game/issues/1300.))

> Also, have you ever noticed that a human can't sink in lava, not even his dead body?

https://github.com/minetest/minetest_game/issues/2296

**Comment by Warr1024** [comment]
I recently ran into this issue in NodeCore.  I want to disable sinking, in at least some circumstances (e.g. feet in water but head above) so players are be able to _naively_ tread water.  I tried setting gravity to 0 but it didn't have any effect.  I tried using add_player_velocity but counteracting sinking with jolts of upward buoyancy leads to very jerky motion, and the resulting camera jank was bad enough to make building or digging while floating badly broken and the whole feature had to be removed.  Being able to set a speed (or an acceleration) for sinking/buoyancy would have resolved this perfectly.

**Comment by Hume2** [comment]
If it is all so hard to implement, can you at least scale the sinking speed by the gravity?

That shouldn't require more than changing one or a few lines of code and it would satisfy nearly all the needs.

**Comment by paramat** [comment]
Warr1024,
Have you tried editing `movement_liquid_sink` to be negative? That is the obvious already present solution. However, the player might end up bobbing on the surface, not half submerged as you ideally want.

Hume2,
It is not hard to implement, the MT-physics-consistent and simple solution is explained in https://github.com/minetest/minetest/issues/6123#issuecomment-506135735 . That will implement what was the original request: "sinking in liquids doesn't depend on player's gravity".
Does anyone want to make a PR to test? It should be simple.
> can you at least scale the sinking speed by the gravity?

It makes no sense to scale 'speed'.
> That shouldn't require more than changing one or a few lines of code

You are just guessing.
That is incorrect and you do not understand what you are writing about.

**Comment by Wuzzy2** [comment]
> A liquid_sink_speed field could also be added to the override_table of set_physics_override.

This will be done in https://github.com/minetest/minetest/pull/11465
